### PR TITLE
[WIP] lib: Add herumi/bls to build

### LINKED ISSUE #11: lib: add herumi/mcl and herumi/bls as submodules and to the build process

### PR DESCRIPTION
**Pull request by mxaddict**
closes #11

NOTES: 
 - `libbls384_256.a` is added to the build not no coded currently uses the lib
 - libbls is added as a subtree in `src/bls`, with libmcl also as a subtree inside of `src/bls/mcl`

### CODE REVIEW

**Comment by aguycalled** [comment]
Added this patch to test whether everything was correctly linked:

```diff
diff --git a/src/init.cpp b/src/init.cpp
index f934fd751..63d13ccd5 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -98,6 +98,9 @@
 #include <zmq/zmqrpc.h>
 #endif
 
+#include <mcl/bn_c384_256.h>
+#include <bls/bls.h>
+
 using node::CacheSizes;
 using node::CalculateCacheSizes;
 using node::ChainstateLoadVerifyError;
@@ -1114,6 +1117,20 @@ bool AppInitInterfaces(NodeContext& node)
 
 bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 {
+    blsInit(5, MCLBN_COMPILED_TIME_VAR);
+
+    mclBnG1 g1;
+    mclBnG1_getBasePoint(&g1);
+
+    std::vector<unsigned char> ba(48);
+    int baLen = mclBnG1_serialize(&ba[0], 48, &g1);
+
+    assert(baLen == 48);
+
+    std::cout << "G1 basepoint: " << HexStr(ba);
+
+    exit(-1);
+
     const ArgsManager& args = *Assert(node.args);
     const CChainParams& chainparams = Params();
```

Fails:

```
Undefined symbols for architecture arm64:
  "mcl::fp::Op::init(mcl::VintT<mcl::vint::Buffer<unsigned long long> > const&, unsigned long, int, mcl::fp::Mode, unsigned long)", referenced from:
      mcl::FpT<mcl::bn::local::FpTag, 384ul>::init(bool*, int, mcl::VintT<mcl::vint::Buffer<unsigned long long> > const&, mcl::fp::Mode) in libbls384_256.a(bls_c384_256.o)
      mcl::FpT<mcl::bn::local::FrTag, 256ul>::init(bool*, int, mcl::VintT<mcl::vint::Buffer<unsigned long long> > const&, mcl::fp::Mode) in libbls384_256.a(bls_c384_256.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bitcoind] Error 1
make[2]: *** Waiting for unfinished jobs....
```

**Comment by aguycalled** [comment]
Tested with the code I've shared in my previous comment and did a few more tests using other functions from the library, with all of them working successfully.

**Review by aguycalled — APPROVED** [review]

File: src/Makefile.am [hunk 1]
```diff
@@ -15,7 +15,7 @@ AM_LIBTOOLFLAGS = --preserve-dup-deps
 PTHREAD_FLAGS = $(PTHREAD_CFLAGS) $(PTHREAD_LIBS)
 EXTRA_LIBRARIES =
 
-BITCOIN_INCLUDES=-I$(builddir) -I$(srcdir)/$(MINISKETCH_INCLUDE_DIR_INT) -I$(srcdir)/secp256k1/include -I$(srcdir)/$(UNIVALUE_INCLUDE_DIR_INT) $(BDB_CPPFLAGS) $(BOOST_CPPFLAGS) $(LEVELDB_CPPFLAGS)
+BITCOIN_INCLUDES=-I$(builddir) -I$(srcdir)/$(MINISKETCH_INCLUDE_DIR_INT) -I$(srcdir)/secp256k1/include -I$(srcdir)/bls/include -I$(srcdir)/bls/mcl/include -I$(srcdir)/$(UNIVALUE_INCLUDE_DIR_INT) $(BDB_CPPFLAGS) $(BOOST_CPPFLAGS) $(LEVELDB_CPPFLAGS)
```
[the remark this answers is not in the document]
Comment (mxaddict) [reply src/Makefile.am]: I've added and pushed it, we plan on including .cpp files?

[the remark this answers is not in the document]
Comment (aguycalled) [reply src/Makefile.am]: initially nope but I've seen it added by herumi even if xpp is not used

[the remark this answers is not in the document]
Comment (mxaddict) [reply src/Makefile.am]: Makes sense

**Pull request merged**